rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/LocalInfrastructure.java

```diff
@@ -486,9 +513,35 @@ private void decrementNumberOfLostNodesWithLockAndPersist() {
 
     private int getIndexAndIncrementWithLockAndPersist() {
         return setPersistedInfraVariable(() -> {
-            int deployedNodeIndex = (int) this.persistedInfraVariables.get(LAST_NODE_STARTED_INDEX_KEY);
-            this.persistedInfraVariables.put(LAST_NODE_STARTED_INDEX_KEY, deployedNodeIndex + 1);
-            return deployedNodeIndex;
+            // This method aims at finding the first integer index not already used by a current deployment
+            // The algorithm is the following:
+            // set deployedNodeIndexes = [all currently used indexes]
+            // set allIndexes = [0..max(deployedNodeIndexes)+1]
+            // set availableIndexes = allIndexes - deployedNodeIndexes
+            // newIndex = min(availableIntegers)
+            TreeSet<Integer> deployedNodeIndexes = (TreeSet<Integer>) this.persistedInfraVariables.get(LAST_NODE_STARTED_INDEXES_KEY);
+            int maxIndex;
+            if (deployedNodeIndexes.isEmpty()) {
+                maxIndex = -1;
+            } else {
+                maxIndex = deployedNodeIndexes.last();
+            }
+            ContiguousSet<Integer> allIndexes = ContiguousSet.create(Range.closed(0, maxIndex + 1),
+                                                                     DiscreteDomain.integers());
+            TreeSet<Integer> availableIndexes = new TreeSet(Sets.difference(allIndexes, deployedNodeIndexes));
+            int newIndex = availableIndexes.first();
```

Comment:
What happens when availableIndexes is empty ?

Reply:
it cannot be because allIndexes = [0..max(deployedNodeIndexes)+1] so availableIndexes contains at least max(deployedNodeIndexes)+1